Update CONTRIBUTING Guide (#676)
pyvista/pyvista: CONTRIBUTING.md

The finished documentation can be found in the `docs/_build/html` directory.

#### Creating a New Pull Request

Once you have tested your branch locally, create a pull request on
(pyvista GitHub)[https://github.com/pyvista/pyvista] while merging to

> master.  This will automatically run continious integration (CI)

Reviewer: ```suggestion
master.  This will automatically run continuous integration (CI)
```

The finished documentation can be found in the `docs/_build/html` directory.

#### Creating a New Pull Request

Once you have tested your branch locally, push your branch to (pyvista

> GitHub)[https://github.com/pyvista/pyvista] and create a pull request

Reviewer 2: Non-members of pyvista probably can't do that. Above it's written to fork first, in which case you can push to your own repo and PR from there (CI still kicks in, at least did for me). Or if it's really wanted to have branches in the pyvista repo always, the step above to fork isn't necessary.

Reviewer 3: Now asking that they just submit a PR to pyvista while merging to master.

while merging to master.  This will automatically run continious
integration (CI) testing and verify your changes will work across
several platforms.

To ensure someone else reviews your code, at least one other member of

> the pyvista contributiors group must review and verify your code meets

Reviewer 2: ```suggestion
the pyvista contributors group must review and verify your code meets
```

our community's standards.  Once approved, if you have write
permission you may merge the branch.  If you don't have write
permission, the reviewer or someone else with write permission will

> merge the branch and close out the PR.
> 
> Since it may be necessary to merge your branch with the current
> release branch (see below), please do not close out your branch if it

Reviewer 2: ```suggestion
merge the branch and delete the PR branch.

Since it may be necessary to merge your branch with the current
release branch (see below), please do not delete your branch if it
```

is a `fix/` branch.

### Branching Model

This project follows a modified

> [gitflow](https://datasift.github.io/gitflow/IntroducingGitFlow.html)

Reviewer 2: I think the processes described aren't really that close to gitflow - we are squash merging and stable branches instead of a master branch that holds release but not development. I think oneflow would be closer, but still not really the same: https://www.endoflineblog.com/oneflow-a-git-branching-model-and-workflow. I'd personally just drop the references, the explanations here together with looking at what is actually happening on pyvista should be clear enough imo.

Reviewer 3: Agreed.  Dropping links and just starting with the explanation.

branching model.  The purpose of this is to enable rapid development
of features without sacrificing stability.  These are the main
features:

- The `master` branch is the main development branch.  All features,
  patches, and other branches should be merged here.  While all PRs
  should pass all applicable CI checks, this branch may be
  functionally unstable as changes might have introduced unintended
  side-effects or bugs that were not caught through unit testing.
- There will be one or many `release/` branches based on minor
  releases (for example `release/0.24`) which contain a stable version
  of the code base that is also reflected on PyPi/.  Hotfixes from
  `fix/` branches should be merged both to master and to these
  branches.  When necessary to create a new patch release these
  release branches will have their `__version__.py` updated and be
  tagged with a patched semantic version (e.g. `0.24.1`).  This
  triggers CI to push to PyPi, and allow us to rapidly push hotfixes
  for past versions of `pyvista` without having to worry about
  untested features.

> - When a minor release canidate is ready, a new `release` branch will

Reviewer: ```suggestion
- When a minor release candidate is ready, a new `release` branch will
```

>   be created from `master` will be created with the next incremented

Reviewer: ```suggestion
  be created from `master` with the next incremented
```

>   minor version (e.g. `release/0.25.0`), which will be throughly

Reviewer: ```suggestion
  minor version (e.g. `release/0.25.0`), which will be thoroughly
```

>   minor version (e.g. `release/0.25.0`), which will be throughly

Reviewer 2: Also:

```suggestion
  minor version (e.g. `release/0.25`), which will be throughly
```

Reviewer 3: Looks like only one suggestion can be made per line, otherwise I'd commit this suggestion. Changed as requested.

  tested.  When deemed stable, the release branch will be tagged with
  the version (`0.25.0` in this case), and if necessary merged with
  master if any changes were pushed to it.  Feature development then
  continues on `master` and any hotfixes will now be merged with this

>   release.  Older release branches may be deleted when needed.

Reviewer 2: I'd never delete a release branch. Sure, tags will stay to indicate past releases, but it's still a bit more obscure if they are on commits that aren't connected to any branches.

Reviewer 3: Replaced last sentence with `Older release branches should not be deleted so they can be patched as needed.`

#### Minor Release Steps

Minor releases are feature and bug releases that improve the
functionality and stability of `pyvista`.  Before a minor release is
created the following will occur:

1.  Create a new branch from the `master` branch with name
    `release/MAJOR.MINOR` (e.g. `release/0.25`).

> 2. Locally verify run all tests as outlined in the [Testing Section](#testing) 

Reviewer: ```suggestion
2. Locally run all tests as outlined in the [Testing Section](#testing) 
```

    ```

4. After building the documentation, open the local build and examine the
examples gallery for any obvious issues.

5. Update the version numbers in `pyvista/_version.py` and commit it.
   Push the branch to GitHub and create a new PR for this release that
   merges it to master.  Development to master should be limited at
   this point while effort is focused on the release.

6. It is now the responsibility of the `pyvista` community to
   functionally test the new release.  It is best to locally install
   this branch and use it in production.  Any bugs identified should
   have their hotfixes pushed to this release branch.

7. When the branch is deemed as stable for public release, the PR will
   be merged to master and the `master` branch will be tagged with a

>    `MAJOR.MINOR.0` release.  The release branch will not be deleted.

Reviewer: Tiny question here: so the next minor release will be `0.25.0`?

Reviewer 3: Correct.
